### ISSUE TITLE
🧩 Plugins

### ISSUE THREAD

**Issue by jodydonetti** [jodydonetti/ZiggyCreatures.FusionCache#15]
## Scenario

While playing with the implementation of the backplane (see #11) and talking about adding metrics (see #9) the need emerged to be able to add functionalities around FusionCache via external code.

In some cases a specific interface is needed because it is a core part of the system (like the already existing `IFusionCacheSerializer`), but in a lot of other cases a more generic one would probably suffice.

The objective is to:

- let other people extend FusionCache with custom needs for their projects
- open the door to more contributions by the community for common functionalities not present in the core packages

## Proposal

The idea is to create a **plugin subsystem** where multiple plugins can be implemented using a common interface that would allow the coordination with a FusionCache instance.

As a first draft I'm thinking about something like this:

```csharp
public interface IFusionCachePlugin
{
    void Start(IFusionCache cache);
    void Stop(IFusionCache cache);
}
```

In the `Start` method a plugin implementer will receive a cache instance to then, for example, subscribe to the events they'd like (see #14) or do something else, while in the `Stop` they can remove the events subscriptions they've created before, to keep a system clean.

A plugin will be added to a cache with a method like `IFusioCache.AddPlugin(IFusionCachePlugin plugin)`, similarly to [the one for the distributed cache](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/4dc48a1aa8099d11de7a0f6bf91e537b28771f91/src/ZiggyCreatures.FusionCache/IFusionCache.cs#L26) which would add the plugin instance to an internal list of plugins, and call its `Start` method. In the same vein, a method `IFusioCache.RemovePlugin(IFusionCachePlugin plugin)` can be called to remove a plugin (which in turn will call the `Stop` method on the plugin) for housekeeping purposes.

## Dependency Injection

In a DI scenario the method will be called automatically for all the registered services implementing the IFusionCachePlugin type, like what is already happening for the IDistributedCache type [here](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/4dc48a1aa8099d11de7a0f6bf91e537b28771f91/src/ZiggyCreatures.FusionCache/FusionCacheServiceCollectionExtensions.cs#L57), but with potentially multiple implementations.

The code may be something like this:

```csharp
[...]
var plugins = serviceProvider.GetServices<IFusionCachePlugin>();
foreach (var plugin in plugins)
{
    cache.AddPlugin(plugin);
}
[...]
```

## Specialized plugin types

In case specific functionalities may be needed by FusionCache, a more specialized plugin type may be created simply inheriting from the base `IFusionCachePlugin` type and adding the specific apis needed.

If, for example, the metrics plugins need a special metrics-related method to be called for metrics-related things, we would have something like this:

```csharp
public interface IFusionCacheMetricsPlugin:
    : IFusionCachePlugin
{
    Task<bool> DoMetricsStuffAsync(int param1, string param2, [etc...]);
}
```

**NOTE:** I'm still playing with the backplane impl, and it could very well fit into this as a "normal" plugin.

## Other stuff to decide

Should there be a way to identify a plugin, apart from its clr type? Something like a `string Id { get; set; }`? It may be useful in some contexts (eg: logging).

## Thoughts?

Any suggestion is more than welcome.

**Comment by jodydonetti** [comment]
I just published a new branch with the **plugins** subsystem [here](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/tree/feature_plugins).

## How to create a plugin

Write a class that implements the [`IFusionCachePlugin`](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/feature_plugins/src/ZiggyCreatures.FusionCache/Plugins/IFusionCachePlugin.cs) interface, basically just the `Start` and `Stop` methods, like [this](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/feature_plugins/tests/ZiggyCreatures.FusionCache.Tests/PluginsTests.cs#L13).

Of course it can also accept its own set of options, typically modelled via `IOptions<MyPluginType>` and friends.


## How to use a plugin via DI

Register your plugin type in the DI container (better before registering FusionCache itself) to respond to the `IFusionCachePlugin` service as a singleton, like [here](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/feature_plugins/tests/ZiggyCreatures.FusionCache.Tests/PluginsTests.cs#L197): when a FusionCache instance will be requested, all plugin types registered will be auto-discovered, added (and started) automatically.

Of course you can also define your own custom ext method (like [this one](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/feature_plugins/src/ZiggyCreatures.FusionCache.Serialization.NewtonsoftJson/FusionCacheNewtonsoftJsonSerializerServiceCollectionExtensions.cs#L20)) to be a better .NET citizen.

Example:

```csharp
[...]
services.AddSingleton<IFusionCachePlugin, MyPlugin>();
services.AddFusionCache();
[...]
```

or with a custom ext method + custom options:

```csharp
[...]
services.AddMyFusionCachePlugin(options => {
  options.Whatever = 42;
});
services.AddFusionCache();
[...]
```


## How to use a plugin without DI
Just create an instance of your plugin and pass it to the `cache.AddPlugin` method, like [here](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/feature_plugins/tests/ZiggyCreatures.FusionCache.Tests/PluginsTests.cs#L150).

Remember that, in case the `Start` method of your plugin throws and exception, an `InvalidOperationException` would also be thrown by the `AddPlugin` method itself (with the original exception as the inner one).


## Removing a plugin

If you need continue using a FusionCache instance but want to remove a plugin you've previously added, you can call the `cache.RemovePlugin` method: it will automatically call the `Stop` method of the plugin and then remove it.

If instead you just want to clean things up "at the end" (and you should do it), you don't have to do anything because when the `FusionCache` instance will be disposed, all added plugins will be stopped and removed automatically.


## Opinions?

Anything would be appreciated, thanks!

**Comment by jodydonetti** [comment]
Pinging @JoeShook because of course 😄

**Comment by jodydonetti** [comment]
The plugins subsystem has been released with [v0.1.5](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/releases/tag/v0.1.5) 🎉
